draft-ietf-quic-transport.md

```diff
@@ -2887,6 +2964,94 @@ by a client in protected packets, because it is certain that the server is able
 to decipher the packet.
 
 
+## ACK_ECN Frame {#frame-ack-ecn}
+
+A QUIC connection MUST keep counters for each ECN codepoint, recording the
+number of packets that were received with the corresponding ECN codepoint in the
+IP header. If the header is not readable from the application, the codepoint 00
+(Not-ECT) MUST be assumed. If any packet are duplicated by the network then only
+the value of the ECN field of the packet copy first received SHALL be included
+in the counters. This to prevent the on-side attack ({{security-ecn}}) and
+ensure that ACK_ECN frames becomes idempotent in the event of packet
+duplication. Note, a receiver is not required to maintain indefinite state for
+which packet numbers have been received far into the history. Packets discarded
+for this reason, their ECN values are also not counted.
+
+ACK_ECN Frame MUST be used when when an endpoint is acknowledging a packet were
+the IP header ECN field was marked as ECT(0), ECT(1), or ECN-CE when received.
+ACK Frames ({{frame-ack}}) MAY be sent when all packets to be acknowledged
+had an IP header with the ECN field marked as Not-ECT.
+
+The ACK_ECN frame is used by the receiver to echo the value of these counters
+back to the sender of these packets. This allows the sender to utilize these
+counter values for congestion control. The ACK_ECN frame contains all the
+elements of the ACK frame ({{frame-ack}}) with the addition of an ECN block
+appended at the end.
+
+
+~~~
+ 0                   1                   2                   3
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                     Largest Acknowledged (i)                ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                          ACK Delay (i)                      ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                       ACK Block Count (i)                   ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                          ACK Blocks (*)                     ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                           ECN Block                         ...
```

Comment:
As discussed on the list, I think it makes sense to but the ACK Block Count and ACK blocks at the end of this frame, to make it simpler to truncate the ack blocks if necessary.

Reply:
Agree, if no ones protests, I will implement that change tomorrow.

Reply:
Now implemented